key-share/src/trusted_dealer.rs

```diff
@@ -0,0 +1,182 @@
+            n,
+            shared_secret_key: None,
+            #[cfg(feature = "hd-wallets")]
+            enable_hd: false,
+        }
+    }
+
+    /// Sets threshold value
+    ///
+    /// If threshold is `Some(_)`, resulting key shares will be generated
+    /// using t-out-of-n VSS scheme. If it's `None`, trusted dealer will
+    /// generate additive key shares in n-out-ouf-n scheme.
+    ///
+    /// Note that setting `t=Some(n)` is not the same as setting `t=None`.
+    /// Both produce n-out-of-n key shares, but `t=Some(n)` mocks threshold
+    /// key generation with `threshold=n`, `t=None` mock non-threshold key
+    /// generation.
+    ///
+    /// Default: `None`
+    pub fn set_threshold(self, t: Option<u16>) -> Self {
+        Self { t, ..self }
+    }
+
+    /// Sets shared secret key to be generated
+    ///
+    /// Resulting key shares will share specified secret key.
+    pub fn set_shared_secret_key(self, sk: SecretScalar<E>) -> Self {
+        Self {
+            shared_secret_key: Some(sk),
+            ..self
+        }
+    }
+
+    /// Specifies that the key being generated shall support HD derivation
+    #[cfg(feature = "hd-wallets")]
+    pub fn hd_wallet(self, v: bool) -> Self {
+        Self {
+            enable_hd: v,
+            ..self
+        }
+    }
+
+    /// Generates [`CoreKeyShare`]s
+    ///
+    /// Returns error if provided inputs are invalid, or if internal
+    /// error has occurred.
+    pub fn generate_shares(
+        self,
+        rng: &mut (impl rand_core::RngCore + rand_core::CryptoRng),
+    ) -> Result<Vec<CoreKeyShare<E>>, TrustedDealerError> {
+        let shared_secret_key = self
+            .shared_secret_key
+            .unwrap_or_else(|| SecretScalar::random(rng));
+        let key_shares_indexes = (1..=self.n)
+            .map(|i| generic_ec::NonZero::from_scalar(Scalar::from(i)))
+            .collect::<Option<Vec<_>>>()
+            .ok_or(Reason::DeriveKeyShareIndex)?;
+        let (shared_public_key, secret_shares) = if let Some(t) = self.t {
+            let f = generic_ec_zkp::polynomial::Polynomial::sample_with_const_term(
+                rng,
+                usize::from(t) - 1,
+                shared_secret_key,
+            );
+            let pk = Point::generator() * f.value::<_, Scalar<_>>(&Scalar::zero());
+            let shares = key_shares_indexes
+                .iter()
+                .map(|I_i| f.value(I_i))
+                .map(|mut x_i| SecretScalar::new(&mut x_i))
+                .collect::<Vec<_>>();
+            (pk, shares)
+        } else {
+            let mut shares = std::iter::repeat_with(|| SecretScalar::<E>::random(rng))
+                .take((self.n - 1).into())
+                .collect::<Vec<_>>();
+            shares.push(SecretScalar::new(
+                &mut (shared_secret_key - shares.iter().sum::<Scalar<E>>()),
+            ));
+            let pk = shares.iter().map(|x_j| Point::generator() * x_j).sum();
```

Comment:
I've just realized that it'll be more efficient to sum all the shares, and only then to multiply by G